DOCS-1377: Tutorial for facial verification security system module (#2278)
viamrobotics/docs: docs/tutorials/projects/verification-system.md

---
title: "Create a Facial Verification System"
linkTitle: "Verification System"

> weight: 40

Author: ```suggestion
```

type: "docs"
description: "Configure a verification system module to implement a facial verification security system on your smart machine."
image: "/tutorials/verification-system/disarmed.png"
imageAlt: "Bijan disarming the facial verification system."
images: ["/tutorials/verification-system/disarmed.png"]
tags: ["mlmodel", "vision", "services", "security", "canera"]

> authors: []

Author: ```suggestion
authors: ["Sierra G."]
```

languages: []
viamresources: ["mlmodel", "vision", "camera"]
level: "Intermediate"
date: "28 November 2023"
cost: "0"
no_list: true
# SMEs: Bijan Haney
---

In the [registry](https://app.viam.com/module/viam-labs/verification-system), Viam Labs provides a model of [vision service classifier](/ml/vision/classification/) called `verification-system`.
Follow this tutorial to learn how to configure this module and set up your own security system to detect, verify, and alarm based on specified detections from nested [vision service detectors](/ml/vision/detection/).

> You can use whatever combination of detectors and camera you would like with this module, but follow along with our recommendations to layer an [`mlmodel` person detector](#configure-an-mlmodel-person-detector) and [facial detector](#configure-a-facial-detector) together, creating a security system with facial recognition.

Reviewer: This currently starts off in a way that only someone familiar with the product would understand what is being built. 

Let's try to make this a bit more like a sample project someone built not to show off the module but because they wanted to build the project. Let's start by spelling out what it is we're building and what the core functionality is (how it can recognize a face and that disarms the system etc). Being more specific with this helps the reader understand what's being built. We also should get a picture of the built project - you can spend some time building it. Or maybe take a picture of Bijan's. Cardboard and tape is fine to hold things together if need be. But let's help the audience visualise what is being built. The video of the disarming would also help explain the functionality.

{{< alert title="Requirements" color="note" >}}
Before configuring your camera, you must [create a robot](/fleet/machines/#add-a-new-robot).

> {{< /alert >}}

Reviewer: Let's guide them through that. Also let's list the hardware they need

## Configure a camera

Navigate to the **Config** tab of your machine's page on the [Viam app](https://app.viam.com).

> Configure the camera you want to provide an image stream for your verification system on.

Reviewer: This sounds a bit stilted. You can just say the camera you'll use for your security system or something like that.

Author: Ok, done!

We configured ours as a webcam named `"my-webcam"`.
Follow [these instructions](/build/configure/components/camera/webcam/) to do so.

## Configure an `mlmodel` person detector

Create a detector to detect when people are in your verification system camera's field of vision:

> - [Capture images from your camera](/data/capture/#configure-data-capture-for-individual-components) with Viam's data management service, then [create a dataset](/data/dataset/#create-a-dataset-and-add-data).

Reviewer: Use ordered listed when steps need to be done in a specific order

Author: Done!

- Label your data with [bounding boxes](/data/dataset/#bounding-boxes) labeled as `Person` where a person is present to create an object detection model for people.
- [Train a model on your dataset](/ml/train-model/).

> - [Deploy the model](/ml/deploy/) to your machine.

Reviewer: A little more detail on each step would be helpful . You can explain that you need to take images of the people who should be able to disarm the system or just people (?) and so on

Author: Ok!

Now that you've trained and deployed the model itself, [configure an `mlmodel` detector model](/ml/vision/detection/#configure-an-mlmodel-detector) with the name of the ML model service you deployed your model as.
Name this detector `"people-detect"`.

For example, we configured a person detector as follows:

```json {class="line-numbers linkable-line-numbers"}
"services": [
  {
    "type": "mlmodel",
    "model": "tflite_cpu",
    "attributes": {
      "model_path": "${packages.persondetect0tflite}/persondetect0.tflite",
      "label_path": "${packages.persondetect0tflite}/persondetectlabels.txt",
      "num_threads": 1
    },
    "name": "persondetect"
  },
  {
    "name": "people-detect",
    "type": "vision",
    "model": "mlmodel",
    "attributes": {
      "mlmodel_name": "persondetect"
    }
  }
]
```

> ## Configure a facial detector

Reviewer: We never explained why we're doing this or the people detector

To create a detector that can recognize individual faces, use Viam Lab's `facial-detector` module, available in the [registry](https://app.viam.com/module/viam-labs/facial-detector).
This is a modular vision service that uses the DeepFace library to perform facial detections.

- Navigate to the **Config** tab of your robot’s page in [the Viam app](https://app.viam.com/).
  Click on the **Services** subtab and click **Create service**.
  Select the `vision` type, then select the `detector:facial-detector` model. Enter a name for your vision service and click **Create**.
- Edit the attributes as applicable according to the configuration information on [GitHub](https://github.com/viam-labs/facial-detection). Label a photo of the face of each person you want your security system to recognize in `"face_labels"`.

Name this detector `"face-detect"`.

> For example, we configured a facial detector as follows:

Reviewer: I think you'll need to explain the attributes. I have no idea what those are for and whether they relate to the model i trained and deployed. You don't need to recreate the readme docs but a little more info on what's happening is needed

Author: Added more info

To use the facial detector as the verification detector and the people detector as the first two layers of detectors, with **Builder** mode selected copy and paste the following attributes configuration into your service's **Attributes** box:

```json {class="line-numbers linkable-line-numbers"}
{
  "trigger_1_confidence": 0.35,
  "verification_detector": "face-detect",
  "camera_name": "my-webcam",
  "trigger_2_confidence": 0.5,
  "trigger_1_labels": ["Person"],
  "trigger_2_labels": ["Person"],
  "disable_alarm": false,
  "trigger_2_detector": "people-detect",
  "verification_labels": ["my_name"],
  "trigger_1_detector": "people-detect",
  "disarmed_time_s": 10,
  "countdown_time_s": 10
}
```

Name your service `"security"`.

- As the `"people-detect"` detector serves as both trigger detector 1 and 2, set `"trigger_1_labels"` and `"trigger_2_labels"` to `"Person"` to match the bounding box label of `"Person"` you assigned to images in your dataset when training the object detection model behind this detector.

> - As the verification detector is your facial detector `"face-detect"`, set `"verification_labels"` to match the name(s) you assigned to images in the facial detector modules' `"face_labels"` configuration attribute.

Author: ```suggestion
- As the verification detector is your facial detector `"face-detect"`, set `"verification_labels"` to match each name you assigned to an image in the facial detector modules' `"face_labels"` configuration attribute.
```

Click on the **Components** subtab and click **Create component**. Select the `camera` type, then select the `transform` model. Enter a name for your camera and click **Create**.

With **Builder** mode selected, copy and paste the following attributes JSON into your camera's **Attributes** box:

```json {class="line-numbers linkable-line-numbers"}
{
  "pipeline": [
    {
      "type": "classifications",
      "attributes": {
        "classifier_name": "security",
        "confidence_threshold": 0.5
      }
    }
  ],
  "source": "my-webcam"
}
```

Note that the camera name in `"source"` and the `"classifier_name"` match the names that we configured for our source camera and verification system service, respectively.
If you used different names, edit as applicable.
View these [instructions to configure a transform camera](/build/configure/components/camera/transform/) for reference.
Save your config.

## View your verification system in action

Navigate to the **Control** tab.
Expand the card matching the name of the transform camera you configured to view an image stream with the system's notifications overlaid.

> With this configuration of detectors, your verification system will trigger a countdown when it detects people and disarm the alarm if it detects your face with the facial detector within the countdown period.

Reviewer: This is the kind of info missing in the intro

Author: I moved this to the intro-- do you think that's ok? Seems weird to have the video in twice otherwise I would leave here as well

---
title: "Create a Facial Verification System"
linkTitle: "Verification System"
type: "docs"

> description: "Configure a verification system module to implement a facial verification security system on your smart machine."

Author: Done

image: "/tutorials/verification-system/disarmed.png"
imageAlt: "Bijan disarming the facial verification system."

> images: ["/tutorials/verification-system/disarmed.png"]

Author: Done

tags: ["mlmodel", "vision", "services", "security", "camera"]
authors: ["Sierra G."]
languages: []
viamresources: ["mlmodel", "vision", "camera"]
level: "Intermediate"
date: "28 November 2023"
cost: "0"
no_list: true
# SMEs: Bijan Haney
---

> Follow this tutorial to create your own machine learning (ML) enabled security system, or, "verification system."
> The verification system configuration you will use in this project combines two detectors: The first detector detects whether there are people present, thanks to a person-detection ML model you train yourself.
> You give the second detector photos of known faces and it tells you if a person is recognized.
> 
> The verification system has three possible layers of detectors.
> You layer them together so that objects detected in the images your verification's system camera captures point to countdown, alarmed, or disarmed states.
> With our recommended combination of detectors, your verification system will trigger a countdown when it detects people and disarm the alarm if it detects your face with the facial detector within the countdown period.
> 
> For example:
> 
> {{<video webm_src="/tutorials/verification-system/demo.webm" mp4_src="/tutorials/verification-system/demo.mp4" alt="Bijan interacting with the verification system" poster="/tutorials/verification-system/trigger_1.png">}}

Reviewer: This is a lot better but I think the first sentence really needs to explain the system and I don't think anyone really knows what a verification system is. The rest are fairly minor changes and removing the thing about states which we haven't mentioned yet and the three possible layers. I don't understand that but it's also not explained further - Is that a type?? 

```suggestion
Create a machine learning (ML) enabled security system, or "verification system", that can detect people and verify that the people are meant to be there with facial recognition.
The verification system you will create in this project combines two detectors: 

1. The first detector detects whether there are people present, thanks to a person-detection ML model you train yourself.
2. The second detector verifies that the people present are known.
   You will use an existing model and provide it with a set of photos of known faces.

Layering these two detectors, your verification system will trigger a countdown when it detects people and disarm the alarm if it detects your face with the facial detector within the countdown period.

For example:

{{<video webm_src="/tutorials/verification-system/demo.webm" mp4_src="/tutorials/verification-system/demo.mp4" alt="Bijan interacting with the verification system" poster="/tutorials/verification-system/trigger_1.png">}}

This tutorial uses two layers of detectors to detect people and then faces but if your use case is different, you can adapt the detectors or the layering of the detectors.
```

Author: This suggestion looks great, thank you!

## Requirements

- A [camera](/components/camera/)
- A computer running `viam-server` that the camera can connect to, like a [single-board computer (SBC)](/components/board/)

Before configuring your camera, you must [create a robot](/fleet/machines/#add-a-new-robot).

1. On the locations page of the [Viam app](https://app.viam.com), add a new robot by providing a name in the **New Robot** field and clicking **Add robot**.
2. Click on that robot's name to go to its page.

## Configure a camera

Navigate to the **Config** tab of your machine's page on the [Viam app](https://app.viam.com).
Configure the camera you want to use for your security system.

We configured ours as a webcam named `"my-webcam"`.

> Follow [these instructions](/components/camera/webcam/) to do so.

Reviewer: Please copy the relevant bits here so that folks don't have to jump around the docs?

Author: I don't know if it's wise to put the full attribute table in as it's pretty huge with optional attributes, but added more stepwise information.

Connect your camera to your machine running `viam-server`, and hang it up somewhere where it can see people in your home:

{{<imgproc src="/tutorials/verification-system/camera.jpeg" resize="500x" declaredimensions=true alt="Camera hanging up in office.">}}

Next, configure the person detector, or, the "coarser" layer of the security system that verifies that there's a person moving.

## Configure an `mlmodel` person detector

Create a detector to detect when people are in your verification system camera's field of vision:

> 1. [Capture images from your camera](/data/capture/#configure-data-capture-for-individual-components) with Viam's data management service, then [create a dataset](/data/dataset/#create-a-dataset-and-add-data).

Reviewer: This needs more copy because it's very high level telling people what to do. 

```suggestion
To be able to detect people with your verification system, you first need to train a model that detects people and for that you need data.
You can collect this data using the data management service and your already installed camera.

(Add steps to add the data management service and configure capture, then instruct them to turn capturing on for a while while letting different people go past the camera, then instruct them to turn off data capture so that it doesn't cost them a lot of money. ALTERNATIVELY, we could instruct them to use an off the shelf person detector. I think giving both options would be nice)

1. 
you need to [Capture images from your camera](/data/capture/#configure-data-capture-for-individual-components) with Viam's data management service, then [create a dataset](/data/dataset/#create-a-dataset-and-add-data).
```

Author: Maybe google is just not good, but I haven't been able to find a good person detecting TF model online. Object detection, sure, but not specifically people. Vision team is making one that should be really good re. this scope https://viam.atlassian.net/browse/RSDK-5629 want me to create a ticket to track and add to this?

2. Label your data with [bounding boxes](/data/dataset/#bounding-boxes) labeled as `Person` where a person is present to create an object detection model for people. You only want this model to be able to distinguish between what is and isn't a person, so you can conduct this training step with anyone.
3. [Train a model on your dataset](/ml/train-model/). Select **Model Type** as **Object Detection**.
4. [Deploy the model](/ml/deploy/) to your machine so you can use it.

> Now that you've trained and deployed the model itself, [configure an `mlmodel` detector model](/ml/vision/detection/#configure-an-mlmodel-detector) with the name of the ML model service you deployed your model as.

Reviewer: Here I think you need to add more detailed steps as well.  You put multiple steps in one sentence that people have to figure out for themselves, instead of really guiding the reader

Author: Done

Name this detector `"people-detect"`.

For example, we configured a person detector as follows:

```json {class="line-numbers linkable-line-numbers"}
"services": [
  {
    "type": "mlmodel",
    "model": "tflite_cpu",
    "attributes": {
      "model_path": "${packages.persondetect0tflite}/persondetect0.tflite",
      "label_path": "${packages.persondetect0tflite}/persondetectlabels.txt",
      "num_threads": 1
    },
    "name": "persondetect"
  },
  {
    "name": "people-detect",
    "type": "vision",
    "model": "mlmodel",
    "attributes": {
      "mlmodel_name": "persondetect"
    }
  }
]
```

> This will serve as the first and second layers of your verification system.

Reviewer: How so? Isn't the second layer detecting faces? This needs explanation

> Now, move on to configuring the "finer" layer of the detectors-- the facial recognition detector.

Reviewer: ```suggestion
Now, move on to configuring the "finer" layer of the detectors: the facial recognition detector.
```

## Configure a facial detector

> Now, to disarm our alarm, we want our verification system to be able to give someone an "ok" based on their ID.

Reviewer: So for this to make sense you need to explain that the module that we will be using uses the first detector to start a countdown to an alarm. You did in the introduction but that's a long time ago. It may also make sense to explain that we will add a module that contains the code that runs the alarm system based on the given detectors before we dive into creating the detectors.
A diagram with the module and the component and services would be very helpful

Author: done / made ticket

Edit the attributes as applicable according to the configuration information on [GitHub](https://github.com/viam-labs/facial-detection):

- `"face_labels"`: Label a photo of the face of each person you want your security system to recognize with the name you want for the label paired with the image path.
- `"recognition_model"`: The model to use for facial recognition. `"ArcFace"` is chosen as the default for a good balance of speed and accuracy. See [GitHub](https://github.com/viam-labs/facial-detection) for more options.
- `"detection_framework"`: The detection framework to use for facial detection. `"ssd"` is chosen as the default for a good balance of speed and accuracy.

```json {class="line-numbers linkable-line-numbers"}
{
  "face_labels": {
    "my_name": "/home/me/my-photo.jpg"
  },
  "recognition_model": "ArcFace",
  "detection_framework": "ssd"
}
```

Make sure to name this detector `"face-detect"`.

## Configure a verification system

Now that you've configured your camera, a rough `"people-detect"` detector and a more fine-tuned `"face-detect"` facial detector, you can set up your facial verification system module.

> Navigate to the **Config** tab of your robot’s page in [the Viam app](https://app.viam.com/). Click on the **Services** subtab and click **Create service**. Select the `vision` type, then select the `classifier:verification-system` model. Enter a name for your vision service and click **Create**.

Reviewer: Please add more context for what adding this does (I go a bit into that in the above comments - it's adding the code that makes all of this work)

Author: Added above

To use the facial detector as the verification detector and the people detector as the first two layers of detectors, with **Builder** mode selected copy and paste the following attributes configuration into your service's **Attributes** box:

```json {class="line-numbers linkable-line-numbers"}
{
  "trigger_1_confidence": 0.35,
  "verification_detector": "face-detect",
  "camera_name": "my-webcam",
  "trigger_2_confidence": 0.5,
  "trigger_1_labels": ["Person"],
  "trigger_2_labels": ["Person"],
  "disable_alarm": false,
  "trigger_2_detector": "people-detect",
  "verification_labels": ["my_name"],
  "trigger_1_detector": "people-detect",
  "disarmed_time_s": 10,
  "countdown_time_s": 10
}
```

Name your service `"security"`.

> - As the `"people-detect"` detector serves as both trigger detector 1 and 2, set `"trigger_1_labels"` and `"trigger_2_labels"` to `"Person"` to match the bounding box label of `"Person"` you assigned to images in your dataset when training the object detection model behind this detector.

Reviewer: What are these trigger detector 1 and 2? That's why we need to start with giving context for the module and what it expects.

> - As the verification detector is your facial detector `"face-detect"`, set `"verification_labels"` to match each name you assigned to an image in the facial detector modules' `"face_labels"` configuration attribute.

Reviewer: "As the verification detector is your facial detector" - this presupposes that you've explained what a verification detector is and what it's used for and that it will be the facial detector and I really do not think that is clear at this point.

Author: Added more info about identity verification in the facial verification section

Click on the **Components** subtab and click **Create component**. Select the `camera` type, then select the `transform` model. Enter a name for your camera and click **Create**.

With **Builder** mode selected, copy and paste the following attributes JSON into your camera's **Attributes** box:

```json {class="line-numbers linkable-line-numbers"}
{
  "pipeline": [
    {
      "type": "classifications",
      "attributes": {
        "classifier_name": "security",
        "confidence_threshold": 0.5
      }
    }
  ],
  "source": "my-webcam"
}
```

Note that the camera name in `"source"` and the `"classifier_name"` match the names that we configured for our source camera and verification system service, respectively.
If you used different names, edit as applicable.
View these [instructions to configure a transform camera](/components/camera/transform/) for reference.
Save your config.

## View your verification system in action

Navigate to the **Control** tab.
Expand the card matching the name of the transform camera you configured to view an image stream with the system's notifications overlaid.

> 

Reviewer: Add a screenshot here please

Author: Done

> ## How the verification system works

Reviewer: This probably needs to go futher up possibly after the introduction. Note that this does need a bit of "glue", the states in and of itself aren't really an explanation - you'll want to give an overview of what the module is and what we need to provide it with (config-wise) and how it uses that with the states.

Reviewer: ^ this is the information you have about the three layers you could use that you reference in the intro that the reader will be very confused about

Author: Moved

3. `COUNTDOWN`: This state runs the `verification_detector` on every frame, looking for detections with any label from `verification_labels` with at least `verification_confidence`.
   If the detector triggers, then the state moves to `DISARMED`.
   If it doesn't trigger in the time specified by `countdown_time_s`, it moves to state `ALARM`.
4. `ALARM`: The alarm state.
   The module will emit the `ALARM` classification for the amount of time specified in `alarm_time_s`.
   After that amount of time elapses, the module will return to state `TRIGGER_1`.
5. `DISARMED`: The disarmed state.
   The module will emit the `DISARMED` classification for the amount of time specified in `disarmed_time_s`.
   After that amount of time elapses, the module will return to state `TRIGGER_1`.

If you do not want the `ALARM` capabilities, and would like to just use it as a notification system when a detector gets triggered, you can set `disable_alarm: true` in the config, which will prevent `TRIGGER_2` from entering into the `COUNTDOWN` state.
This means the system will only cycle between the states of `TRIGGER_1` and `TRIGGER_2`.

> You can use entering into the state `TRIGGER_2` as a way to send notifications.

Reviewer: What does it currently? What does the alarm state do? I suppose nothing, but that's not made clear and left for the reader to guess about.

Author: What trigger 2 state does currently is described above, is that ok? Alarm is emitting classification. I can add in another sentence to make it clear that no sound or anything is triggered

## Next Steps

> - Try changing the type of [detectors](/ml/vision/detection/), using different detectors for the `TRIGGER_1` and `TRIGGER_2` states.

Reviewer: Having two bullets without any context-setting intro is a bit bare. Also I think you really want to tell people to go and set the alarm if that's not done by the code so that the system does something when an alarm is triggered!

Author: Addressing

2. `TRIGGER_2`: This state runs the `trigger_2_detector` on every frame, looking for detections with any label from `trigger_2_labels` with at least `trigger_2_confidence`.
   If the detector triggers, then the state moves to `COUNTDOWN`.
   If it doesn't trigger in 10 frames, it returns to state `TRIGGER_1`.
3. `COUNTDOWN`: This state runs the `verification_detector` on every frame, looking for detections with any label from `verification_labels` with at least `verification_confidence`.
   If the detector triggers, then the state moves to `DISARMED`.
   If it doesn't trigger in the time specified by `countdown_time_s`, it moves to state `ALARM`.
4. `ALARM`: The alarm state.
   The module will emit the `ALARM` classification for the amount of time specified in `alarm_time_s`.
   After that amount of time elapses, the module will return to state `TRIGGER_1`.
5. `DISARMED`: The disarmed state.
   The module will emit the `DISARMED` classification for the amount of time specified in `disarmed_time_s`.
   After that amount of time elapses, the module will return to state `TRIGGER_1`.

> If you do not want the `ALARM` capabilities, and would like to just use it as a notification system when a detector gets triggered, you can set `disable_alarm: true` in the config, which will prevent `TRIGGER_2` from entering into the `COUNTDOWN` state.

Author: ```suggestion
Note that `ALARM` is by default silent and only a label overlay, but you can use the [classifications](/ml/vision/#getclassificationsfromcamera) emitted to set up a physical alarm trigger.
If you do not want the `ALARM` capabilities, and would like to just use it as a notification system when a detector gets triggered, you can set `disable_alarm: true` in the config, which will prevent `TRIGGER_2` from entering into the `COUNTDOWN` state.
```

This means the system will only cycle between the states of `TRIGGER_1` and `TRIGGER_2`.

You can use entering into the state `TRIGGER_2` as a way to send notifications.

## Next Steps

> - Try changing the type of [detectors](/ml/vision/detection/), using different detectors for the `TRIGGER_1` and `TRIGGER_2` states.

Author: ```suggestion
Now that you've got the verification aspect of your system working, you can use this as a launching point for a DIY home security system.
For example:

- Try making it so your machine triggers a physical alarm when the `ALARM` classification is emitted. Write a program with a [Viam SDK](/sdks/) and use [`GetClassificationsFromCamera()`](/ml/vision/#getclassificationsfromcamera) to poll for the classifications made.
- Try changing the type of [detectors](/ml/vision/detection/), using different detectors for the `TRIGGER_1` and `TRIGGER_2` states.
```

   If it doesn't trigger in the time specified by `countdown_time_s`, it moves to state `ALARM`.
4. `ALARM`: The alarm state.
   The module will emit the `ALARM` classification for the amount of time specified in `alarm_time_s`.
   After that amount of time elapses, the module will return to state `TRIGGER_1`.
5. `DISARMED`: The disarmed state.
   The module will emit the `DISARMED` classification for the amount of time specified in `disarmed_time_s`.
   After that amount of time elapses, the module will return to state `TRIGGER_1`.

Note that `ALARM` is by default silent and only a label overlay, but you can [poll the classifications](/ml/vision/#getclassificationsfromcamera) to set up a physical alarm trigger.
If you do not want the `ALARM` capabilities, and would like to just use it as a notification system when a detector gets triggered, you can set `disable_alarm: true` in the config, which will prevent `TRIGGER_2` from entering into the `COUNTDOWN` state.
This means the system will only cycle between the states of `TRIGGER_1` and `TRIGGER_2`.

You can use entering into the state `TRIGGER_2` as a way to send notifications.

## Next Steps

> Now that you've got the verification aspect of your system working, you can use this as a launch point for creating your own DIY home security system.

Author: ```suggestion
Now that you've got the verification aspect of your system working, you can use this as a launch point for customizing your own DIY home security system.
```

---
title: "Create a Facial Verification System"
linkTitle: "Verification System"
type: "docs"
description: "Create an alarm system that can detect people and can recognize faces, allowing it to smartly trigger alarms."
image: "/tutorials/verification-system/disarmed.png"
imageAlt: "Bijan disarming the facial verification system."
images: ["/tutorials/verification-system/demo.gif"]
tags: ["mlmodel", "vision", "services", "security", "camera"]
authors: ["Sierra G."]
languages: []
viamresources: ["mlmodel", "vision", "camera"]
level: "Intermediate"
date: "28 November 2023"
cost: "0"
no_list: true
# SMEs: Bijan Haney
---

Create a machine learning (ML) enabled security system, or "verification system", that can detect people and verify that the people are meant to be there with facial recognition.
The verification system you will create in this project combines two detectors:

1. The first detector detects whether there are people present, thanks to a person-detection ML model you train yourself.
2. The second detector verifies that the people present are known.

>    You will use an existing model and provide it with a set of photos of known faces.

Author: Thank you Andrew, great explanation!

   If it doesn't trigger in the time specified by `countdown_time_s`, it moves to state `ALARM`.
4. `ALARM`: The alarm state.
   The module will emit the `ALARM` classification for the amount of time specified in `alarm_time_s`.
   After that amount of time elapses, the module will return to state `TRIGGER_1`.
   Note that `ALARM` is by default silent and only a label overlay, but you can [poll the classifications](/ml/vision/#getclassificationsfromcamera) to set up an alarm trigger.
5. `DISARMED`: The disarmed state.
   The module will emit the `DISARMED` classification for the amount of time specified in `disarmed_time_s`.
   After that amount of time elapses, the module will return to state `TRIGGER_1`.

Layering these two detectors, your verification system will trigger a countdown when it detects people and disarm the alarm if it detects your face with the facial detector within the countdown period.

For example:

{{<video webm_src="/tutorials/verification-system/demo.webm" mp4_src="/tutorials/verification-system/demo.mp4" alt="Bijan interacting with the verification system" poster="/tutorials/verification-system/trigger_1.png">}}

> This tutorial uses two layers of detectors to detect people and then faces but if your use case is different, you can adapt the detectors or the layering of the detectors.

Reviewer 2: Ah, just noticed this is on the readme. I think definitely better, then, to omit the states in preference to sending the reader to the readme!

## Requirements

- A [camera](/components/camera/)
- A computer running `viam-server` that the camera can connect to, like a [single-board computer (SBC)](/components/board/)

> Before configuring your camera, you must [create a robot](/fleet/machines/#add-a-new-robot).

Author: ```suggestion
Before configuring your camera, you must [create a robot](/fleet/machines/#add-a-new-machine).
```

1. On the locations page of the [Viam app](https://app.viam.com), add a new robot by providing a name in the **New Robot** field and clicking **Add robot**.
2. Click on that robot's name to go to its page.

## Configure a camera

Navigate to the **Config** tab of your machine's page on the [Viam app](https://app.viam.com).
Configure the camera you want to use for your security system.

We configured ours as a `webcam`, but you can use whatever model of camera you'd like.

> Reference [these supported models](/components/camera/#supportedmodels).

Author: ```suggestion
Reference [these supported models](/components/camera/#supported-models).
```

> 1. To configure a `webcam`, nvigate to the **Config** tab of your robot's page in [the Viam app](https://app.viam.com).

Author: ```suggestion
1. To configure a `webcam`, navigate to the **Config** tab of your robot's page in [the Viam app](https://app.viam.com).
```

2. Click on the **Components** subtab and click **Create component**.
3. Select the `camera` type, then select the `webcam` model.

> 4. Enter the name `"my_webcam""` for your camera and click **Create**.

Reviewer 2: EDIT: I see we do this a lot, and elsewhere. Feel free to ignore this comment. Up to you.

Navigate to the **Config** tab of your Machine’s page in [the Viam app](https://app.viam.com/).
Click on the **Services** subtab and click **Create service**.
Select the `vision` type, then select the `detector:facial-detector` model.
Give your vision service the name `"face-detect"` and click **Create**.

Edit the attributes as applicable according to the configuration information on [GitHub](https://github.com/viam-labs/facial-detection):

- `"face_labels"`: Label a photo of the face of each person you want your security system to recognize with the name you want for the label paired with the image path.
- `"recognition_model"`: The model to use for facial recognition. `"ArcFace"` is chosen as the default for a good balance of speed and accuracy. See [GitHub](https://github.com/viam-labs/facial-detection) for more options.
- `"detection_framework"`: The detection framework to use for facial detection. `"ssd"` is chosen as the default for a good balance of speed and accuracy.

```json {class="line-numbers linkable-line-numbers"}
{
  "face_labels": {
    "my_name": "/home/me/my-photo.jpg"
  },
  "recognition_model": "ArcFace",
  "detection_framework": "ssd"
}
```

> Make sure to name this detector `"face-detect"`.

Author: Yeah, I don't know how it would work another way. I added a clarification and a note about SCP with a link to an involved guide

## Configure a verification system

Now that you've configured your camera, a rough `"people-detect"` detector and a more fine-tuned `"face-detect"` facial detector, you can add a module that contains the code that runs the alarm system based on detections it receives from the detectors specified in its configuration.

> Navigate to the **Config** tab of your Machine’s page in [the Viam app](https://app.viam.com/). Click on the **Services** subtab and click **Create service**. Select the `vision` type, then select the `classifier:verification-system` model. Enter a name for your vision service and click **Create**.
> 
> To use the facial detector as the verification detector and the people detector as the first two layers of detectors, with **Builder** mode selected copy and paste the following attributes configuration into your service's **Attributes** box:
> 
> ```json {class="line-numbers linkable-line-numbers"}
> {
>   "trigger_1_confidence": 0.35,
>   "verification_detector": "face-detect",
>   "camera_name": "my-webcam",
>   "trigger_2_confidence": 0.5,
>   "trigger_1_labels": ["Person"],
>   "trigger_2_labels": ["Person"],
>   "disable_alarm": false,
>   "trigger_2_detector": "people-detect",
>   "verification_labels": ["my_name"],
>   "trigger_1_detector": "people-detect",
>   "disarmed_time_s": 10,
>   "countdown_time_s": 10
> }
> ```
> 
> Name your service `"security"`.
> 
> - Trigger detectors 1 and 2 cascade to the alarm state.
>   Since the `"people-detect"` detector serves as both trigger detector 1 and 2, set `"trigger_1_labels"` and `"trigger_2_labels"` to `"Person"` to match the bounding box label of `"Person"` you assigned to images in your dataset when training the object detection model behind this detector.
> - As the identity verification detector is your facial detector `"face-detect"`, set `"verification_labels"` to match each name you assigned to an image in the facial detector modules' `"face_labels"` configuration attribute.
> - Note that `"camera_name": "my_webcam"` matches the name we configured for our source camera.
>   If you used a different name for your camera, change the name in the JSON.
> - Edit the other attributes to reflect your desired confidence thresholds and times between states.

Reviewer 2: Nit: Let's render this as steps also:

````suggestion
1. On your machine's **Config** page in the [Viam app](https://app.viam.com), navigate to the **Services** tab.
1. Click the **Create services** button at the bottom of the page, select `vision`, then select the `[filtered-camera](classifier:verification-system)` model.
   You can also search for `verification-system` directly.
1. Name your modular vision service `security`, then click **Create**.
1. On the panel that appears, enter the following configuration into the **Attributes** field:

   ```json {class="line-numbers linkable-line-numbers"}
   {
     "trigger_1_confidence": 0.35,
     "verification_detector": "face-detect",
     "camera_name": "my-webcam",
     "trigger_2_confidence": 0.5,
     "trigger_1_labels": ["Person"],
     "trigger_2_labels": ["Person"],
     "disable_alarm": false,
     "trigger_2_detector": "people-detect",
     "verification_labels": ["my_name"],
     "trigger_1_detector": "people-detect",
     "disarmed_time_s": 10,
     "countdown_time_s": 10
   }
   ```

   In the configuration above:

   - `"trigger_1_detector"` and `"trigger_2_detector"` both use the `people-detect` ML model you created to determine if a person is present in the camera frame.
      For this tutorial, you are configuring both of these triggers identically to use the person detection ML model. 
   - `"trigger_1_labels"` and `"trigger_2_labels"` similarly both use the `"person"` label you added to images when training the `people-detect` model.
      For this tutorial, you are configuring both of these labels identically to use the person detection ML model. 
   - `"verification_detector"` uses the `face-detect` ML model you configured when you added images of faces to approved and labelled them in the configuration.
   - `"verification_labels"` contains an array of approved names that match each name you assigned to an image in the `facial-detector` modules' `"face_labels"` configuration attribute.
   - `"camera_name"` is the name of the camera to use to detect people and faces.
      If you used a different name for your camera, update this parameter with your camera's name.
   - Edit the other attributes to reflect your desired confidence thresholds and times between states.

See the [`verification-system` module documentation](https://github.com/viam-labs/verification-system) for more information about the trigger states and their various configuration options.
````

Anticipate: users will intuit trigger_1 means person-detect and trigger_2 means face-detect, but this is not so. So I think we definitely need to tell them that, for the purposes of this tutorial, we're not using a more advanced feature of the module whereby we could use 1 and 2 separately. I don't know what that could be, but I'm inferring from this copy and the code that we're not using it here, and that trigger_1 and trigger_2 should be configured identically in this tutorial's conf. I have offered a guess in my above suggestion, but please correct as you know!!

| `trigger_2_labels` | array | **Required** | The class names from `trigger_2_detector` that count as valid. |
| `trigger_2_confidence` | float | Optional | The detection confidence needed in order to move into the `COUNTDOWN` state. <br> Default: `0.5` |
| `verification_detector` | string | **Required** | The name of the vision service detector that you want to use to verify the object. |
| `verification_labels` | array | **Required** | The class names from `verification_detector` that count as valid. |
| `verification_confidence` | float | Optional | The detection confidence needed in order to move into the `DISARMED` state. <br> Default: `0.8` |
| `countdown_time_s` | int | Optional | The time in seconds the system will remain in state `COUNTDOWN` before transitioning to state `ALARM`. <br> Default: `20` |
| `alarm_time_s` | int | Optional | The time in seconds the system will remain in  state `ALARM` before transitioning to state `TRIGGER_1`. <br> Default: `10` |
| `disarmed_time_s` | int | Optional | The time in seconds the system will remain in  state `DISARMED` before transitioning to state `TRIGGER_1`. <br> Default: `10` |

> | `disable_alarm` | bool | Optional | Disables the `COUNTDOWN` and `ALARM` states. The system will always remain in the `TRIGGER_1` and `TRIGGER_2` states. <br> Default: `false` |

Reviewer 2: OPT: Ah, didn't see this table before writing my comment above. You can omit my attribute explanations entirely then, in favor of this table. Or retain if you think helpful to have both specific (this one here means this) and generic (the full list of possibles is this). Up to you!

Reviewer 2: Ah, actually I think given that this is all in the readme, and this tutorial has already discussed the attributes required for this tutorial in lines 218-226, we can omit this entire table. Similar to the states table.

Note that `ALARM` is by default silent and only a label overlay, but you can [poll the classifications](/ml/vision/#getclassificationsfromcamera) to set up a physical alarm trigger.

## Configure a transform camera

> To view the classifications that your verification system makes, configure a transform camera.

Reviewer 2: ```suggestion
At this point, your machine is fully capable of detecting people in its camera feed, and of identifying whether a specific detected person is "approved" (defined under `"face_labels"`) or not.
To easily see this in action, you can add a [transform camera](/components/camera/transform/) to your machine to overlay the current state of the on top of the camera feed.
```

> Navigate to the **Config** tab of your machine’s page in [the Viam app](https://app.viam.com/).
> Click on the **Components** subtab and click **Create component**. Select the `camera` type, then select the `transform` model. Enter a name for your camera and click **Create**.
> 
> With **Builder** mode selected, copy and paste the following attributes JSON into your camera's **Attributes** box:
> 
> ```json {class="line-numbers linkable-line-numbers"}
> {
>   "pipeline": [
>     {
>       "type": "classifications",
>       "attributes": {
>         "classifier_name": "security",
>         "confidence_threshold": 0.5
>       }
>     }
>   ],
>   "source": "my-webcam"
> }
> ```
> 
> Note that the camera name in `"source"` and the `"classifier_name"` match the names that we configured for our source camera and verification system service, respectively.
> If you used different names, edit as applicable.
> View these [instructions to configure a transform camera](/components/camera/transform/) for reference.
> Save your config.

Reviewer 2: Nit: let's convert this to steps as well:

````suggestion
To add a transform camera to your machine:

1. On your machine's **Config** page in the [Viam app](https://app.viam.com), navigate to the **Components** tab.
1. Click the **Create component** button at the bottom of the page, select `camera`, then select the built-in `transform` model.
1. Give the transform camera a name, like `my-transform-camera`, then click **Create**.
1. On the panel that appears, enter the following configuration into the **Attributes** field:

   ```json {class="line-numbers linkable-line-numbers"}
   {
     "pipeline": [
       {
         "attributes": {
           "classifier_name": "security",
           "confidence_threshold": 0.5
         },
         "type": "classification"
       }
     ],
     "source": "my-webcam"
   }
   ```

   If you used different names for the vision service or the camera component, update this configuration with those names.
   You can adjust the `confidence_threshold` to suit your needs.
   A value of `0.5` is a relatively loose match, representing 50% confidence.

1. Click **Save Config** at the bottom of the window to save your changes.
````

## View your verification system in action

> Navigate to the **Control** tab.
> Expand the card matching the name of the transform camera you configured to view an image stream with the system's notifications overlaid:
> 
> ![Verification camera feed](/tutorials/verification-system/disarmed.png)

Reviewer 2: Nit: adding a small intro and converting to steps. Also adding a little reminder of what the red text they're now seeing means, to tie it all together:

```suggestion
With everything configured, you are now ready to see your facial recognition machine in action by watching the transform camera as a person passes in front of the camera.

To view your machine's transform camera overlay:

1. On your machine's **Control** page in the [Viam app](https://app.viam.com), select the transform camera pane, which is listed by the name you gave it in the previous session, such as `my-transform-camera`.
1. Enable the view toggle to see a live camera feed from your camera, overlayed by the current state of the `verification-system` module, which should be `TRIGGER_1` if no people are present in-frame.
1. Have one or more people walk in front of the camera and look directly into it.
   Watch the state change to `COUNTDOWN` and then `DISARMED` when an approved person is detected, or to `ALARM` if no approved person appears within 10 seconds! 

   ![Verification camera feed](/tutorials/verification-system/disarmed.png)
```

## Next Steps

Now that you've got the verification aspect of your system working, you can use this as a launch point for customizing your own DIY home security system.
For example:

> - Try making it so your machine triggers a physical alarm when the `ALARM` classification is emitted. Write a program with a [Viam SDK](/sdks/) and use [`GetClassificationsFromCamera()`](/ml/vision/#getclassificationsfromcamera) to poll for the classifications made.

Reviewer 2: This is the major purpose of this module: to make a sound or light up an LED screen or dial 911 when a particular state is reached. I agree with Naomi that we need more here to bring that home. Here's a rough take:

```suggestion
- Write a program using one of the [Viam SDK](/sdks/) to poll the `facial-verification` module for its current state, and take action when a particular state is reached.
   For example, you could use [`GetClassificationsFromCamera()`](/ml/vision/#getclassificationsfromcamera) to capture when a transition into the `ALARM` state occurs, and then send you an Email with the captured image of the trespasser!
```

- Try changing the type of [detectors](/ml/vision/detection/), using different detectors for the `TRIGGER_1` and `TRIGGER_2` states.

> - Configure the [filtered camera module](https://app.viam.com/module/erh/filtered-camera) and use it as the source camera in your verification system to save images to the Viam cloud when the system enters into specific states.

Reviewer 2: Opt: we could link to our tutorial on it instead, which itself links to the module: https://docs.viam.com/tutorials/projects/filtered-camera/. Up to you.

But definitely mention that only matching images are sent, which is the primary selling point of that module. Maybe:

```suggestion
- Add the [filtered camera module](/tutorials/projects/filtered-camera/) to your machine, and use it as the source camera in your verification system in order to save images to the Viam cloud only when the system enters into specific states.
   This way, you could limit the images captured and synced to only those you are interested in reviewing later, for example.
```

Author: Haha when this was written there was no tutorial, sounds good!

- If you don't want the `ALARM` capabilities, and would like to just use it as a notification system when a detector gets triggered, set `disable_alarm: true` in the config, which prevents `TRIGGER_2` from entering into the `COUNTDOWN` state, meaning the system will only cycle between the states of `TRIGGER_1` and `TRIGGER_2`.

>   - Use entering into the state `TRIGGER_2` as a way to send notifications.

Reviewer 2: ```suggestion
- Use entering into the state `TRIGGER_2` as a way to send notifications.
```

   ```json {class="line-numbers linkable-line-numbers"}
   {
     "pipeline": [
       {
         "attributes": {
           "classifier_name": "security",
           "confidence_threshold": 0.5
         },
         "type": "classification"
       }
     ],
     "source": "my-webcam"
   }
   ```

   If you used different names for the vision service or the camera component, update this configuration with those names.
   You can adjust the `confidence_threshold` to suit your needs.
   A value of `0.5` is a relatively loose match, representing 50% confidence.

1. Click **Save Config** at the bottom of the window to save your changes.

## View your verification system in action

{{% alert title="Note" color="note" %}}
The various states do not cause anything to happen on their own besides appearing as overlays on the transform cam.
To trigger an audio alarm or otherwise have your machine take an action based on the reported state, you can write your own logic using one of the [Viam SDKs](https://docs.viam.com/build/program/) to [poll the classifications](/ml/vision/#getclassificationsfromcamera).

> {{% /alert %}}

Reviewer 2: Hmm, I guess this alert relatively duplicates the first Next Steps. I think it's important enough to be here twice though 👍

Navigate to the **Config** tab of your machine's page on the [Viam app](https://app.viam.com).
Configure the camera you want to use for your security system.

We configured ours as a `webcam`, but you can use whatever model of camera you'd like.
Reference [these supported models](/components/camera/#supported-models).

1. To configure a `webcam`, navigate to the **Config** tab of your Machine's page in [the Viam app](https://app.viam.com).
2. Click on the **Components** subtab and click **Create component**.
3. Select the `camera` type, then select the `webcam` model.
4. Enter the name `my_webcam` for your camera and click **Create**.
5. If your machine is online and connected to the Viam app, your camera’s video path is automatically detected and configured.
   If your machine is not currently connected, you can manually select the video path for your camera, or bring your machine online to have this path automatically configured for you.

Position your camera somewhere where it can easily see the people it will be configured to detect.

{{<imgproc src="/tutorials/verification-system/camera.jpeg" resize="500x" declaredimensions=true alt="Camera hanging up in office.">}}

> Next, configure the person detector, or, the "coarser" layer of the security system that verifies that there's a person moving.

Author: ```suggestion
Next, configure the person detector, or, the coarser layer of the security system that verifies that there's a person moving.
```

3. Choose `Data Management` as the type and specify a name for your data management service, such as `data-manager`.
   Click **Create**.
4. On the panel that appears, you can manage the capturing and syncing functions individually.
   By default, the data management service captures data to the <file>~/.viam/capture</file> directory, and syncs captured data files to the Viam app every 6 seconds (`0.1` minutes in the configuration).
   Leave the default settings as they are, and click **Save Config** at the bottom of the screen to save your changes.
5. Navigate to the **Components** subtab.
   Scroll to the panel of the camera you just configured.
   Find the **Data Capture Configuration** section.
   Click **Add Method**.
   If you're using a webcam, select the method **Type** [`ReadImage()`](/components/camera/#getimage).
   Set the **Frequency** to `0.333`.
   This will capture an image from the camera roughly once every 3 seconds.
   Click **Save config.**
6. Toggle the **Data capture configuration** on.
   Now, your camera is taking pictures.
   Walk in front of it a number of times, perhaps with a friend or two, letting the camera capture many images of you.

>    For optimal performance, try capturing a variety of angles and use different lighting.

Author: ```suggestion
   For best results, try capturing a variety of angles and use different lighting.
```

# SMEs: Bijan Haney
---

Follow this tutorial to create a simple security system with visual people detection to alarm, and facial recognition to disarm.
In this project, you will create a security system with facial recognition to "verify" whether intruders present are recognized by your system and can disable your alarm.

> To accomplish this on your machine, you will configure an [`mlmodel` person detector](#configure-an-mlmodel-person-detector) and [facial detector](#configure-a-facial-detector) together in the [`verification-system` module](https://app.viam.com/module/viam-labs/verification-system).

Reviewer: - I think this goes beyond simple so let's just call it a security system
- please start this in even simpler terms - you're starting the first sentence with alarming and disarming but what does this mean. 
- you use the words "intruders present" but they're just people at that point, if they're not recognized, then they're intruders. This is confusing.

You can try walking the reader through what happens when the system is set up and a person walks into view. What does the machine check and then what does it do.

Author: Rephrased. I think at the last point that's what I'm trying to do on lines 25-26?

> The verification system module itself is a model of [vision service classifier](/ml/vision/classification/), a type of vision service that returns labels for a given image, but it works because of these layered [vision service detectors](/ml/vision/detection/), object detectors which return bounding boxes drawn around the identified objects.

Reviewer: This is a very long sentence and here isn't the right place for mentioning the model types - if I don't already know what they are, I'll probably click on that and get distracted. And you're not connecting what these layers do. This is more the level of explanation you should give here, you can get into the more technical aspects and mention bounding boxes and configuration later.
```suggestion
The verification system module you will use in this project combines two layers of detectors: The first detector detects whether there are people present. The second detector is trained on known faces and can tell you ...
```

Author: Added

> With the combination of detectors you configure on your machine, your verification system will trigger a countdown when it detects people and disarm the alarm if it detects your face with the facial detector within the countdown period.

Reviewer: This sentence needs to merge with the first paragraph. And then try to rewrite that paragraph - start by just explaining the functionality without mentioning any configuration or models: You have a security system with a camera that can detect people and use facial recognition to decide whether they are intruders or not. Go into a little more detail on what happens in each case.

Author: Moved here, and I think added some more info, but really struggling to think of more details to add for this particular case

# SMEs: Bijan Haney
---

In the [registry](https://app.viam.com/module/viam-labs/verification-system), Viam Labs provides a [vision service classifier](/ml/vision/classification/) model called `verification-system`.
Follow this tutorial to learn how to configure this module and set up your own security system to detect, verify, and alarm based on specified detections from nested [vision service detectors](/ml/vision/detection/).

You can use whatever combination of detectors and camera you would like with this module, but follow along with our recommendations to layer an [`mlmodel` person detector](#configure-an-mlmodel-person-detector) and [facial detector](#configure-a-facial-detector) together, creating a security system with facial recognition.

> With this combination of detectors layered together, your verification system will trigger a countdown when it detects people and disarm the alarm if it detects your face with the facial detector within the countdown period.

Reviewer: So this should come first and you should explain more what this project does. Speaking of "whatever combination of detectors" may just about work for someone familiar with what a detector is and how they work but we should provide some help for forming a mental model to understanding what this project is (and then how the module and detectors fit in). You never even cover what a verification system is. And that's not really something we can take for granted people will just know.

Please start the tutorial by covering what the project does and how it is useful (not how the module is useful, how the PROJECT is useful). The module is just a tool to be able to build this project.
